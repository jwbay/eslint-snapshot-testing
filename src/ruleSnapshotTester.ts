import * as fs from 'fs'
import { Linter, Rule } from 'eslint'
import { getFixtureDirectory } from './inferTestDirectory'
import { serializeLintResult } from './lintResultSerializer'
import { findFixtureFile } from './fixtureFinder'
import { parseFixture } from './fixtureParser'

interface Options {
	/**
	 * The rule object, with 'create' and 'meta' properties.
	 *
	 * @example
	 *
	 * runLintFixtureTests({
	 *   ...
	 *   rule: require('../rules/my-custom-rule')
	 * })
	 */
	rule: Rule.RuleModule
	/**
	 * The rule name. Used for test names and to locate fixtures.
	 *
	 * @example
	 *
	 * runLintFixtureTests({
	 *   ...
	 *   rulename: 'my-rule-name'
	 * })
	 */
	ruleName: string
	/**
	 * Used to locate fixtures. An attempt is made to infer the directory of
	 * the calling test, but if this fails, pass `__dirname` here explicitly.
	 *
	 * @example
	 *
	 * runLintFixtureTests({
	 *   ...
	 *   fixtureDirectory: __dirname
	 * })
	 */
	fixtureDirectory?: string
	/**
	 * Use this to set parser options for the rule being tested. These are
	 * permissive by default.
	 *
	 * @example
	 *
	 * runLintFixtureTests({
	 *   ...
	 *   eslintConfig: {
	 *     parserOptions: {
	 *       ecmaVersion: 2020,
	 *       sourceType: 'module',
	 *     }
	 *   }
	 * })
	 */
	eslintConfig?: Omit<Linter.Config, 'rules'>
}

const defaultLintConfig: Linter.Config = {
	parserOptions: {
		ecmaVersion: 2020,
		sourceType: 'module',
		ecmaFeatures: {
			globalReturn: true,
			jsx: true,
		},
	},
}

/**
 * Runs tests for a given rule based on a test fixture located by convention. The fixture
 * should contain a mix of valid and invalid code. Test cases will be generated, and the
 * lint results for the fixture code will be snapshotted by Jest.
 *
 * The test fixture should be named `my-rule-name.fixture`, or `my-rule-name.fixture.{ts,js}`.
 * It should either be in the same directory as the test, or in a `__fixtures__` directory
 * under the test's directory.
 *
 * @example
 * import myCustomRule from '../my-custom-eslint-rule';
 * import { runLintFixtureTests } from 'eslint-snapshot-rule-tester';
 *
 * runLintFixtureTests({
 *   rule: myCustomRule,
 *   ruleName: 'my-custom-rule',
 * });
 */
export function runLintFixtureTests({
	rule,
	ruleName,
	eslintConfig = defaultLintConfig,
	fixtureDirectory = getFixtureDirectory(new Error('getFixtureDirectory')),
}: Options) {
	const fixturePath = findFixtureFile(ruleName, fixtureDirectory)
	const fixtureSource = fs.readFileSync(fixturePath, 'utf8').replace(/\r\n/g, '\n')
	const tests = parseFixture(fixtureSource, fixturePath)
	tests.forEach((entry) => {
		test(entry.testName, () => {
			const linter = new Linter({})
			linter.defineRule(ruleName, rule)
			const result = linter.verify(
				entry.testSource,
				{
					...eslintConfig,
					rules: {
						[ruleName]: entry.ruleOptions ? ['error', ...entry.ruleOptions] : 'error',
					},
				},
				entry.fileName
			)
			const serialized = serializeLintResult({
				lintMessages: result,
				lintedSource: entry.testSource,
			})
			expect(serialized).toMatchSnapshot()
		})
	})
}
