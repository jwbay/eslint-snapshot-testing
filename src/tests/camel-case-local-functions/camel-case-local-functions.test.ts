import { run } from '../ruleSnapshotTester'
import rule from '../../rules/camel-case-local-functions'
import { RuleTester } from 'eslint'

/**
 * .fixture extension under __fixtures__
 *
 * Extract test cases out of comments?
 *      // test: should blah blah blah
 *  sure but make this optional
 * Configuration......... idk
 *
 * keep in mind there will be ONE snap file per rule/test file
 * but potentially MULTIPLE fixtures.... which implies directory organization
 * OR a first class DSL as above to split test cases in fixtures (with config)
 *
 * by default, look for fixture files 1. next to test 2. in __fixtures__
 *
 */

new RuleTester({}).run('wat', rule, {
	valid: [
		{
			code: `
				function foo() {};
				foo();`,
		},
	],
	invalid: [
		{
			code: `
			function Foo() {};
			Foo();`,
			errors: [
				{
					message: 'local function names should be camelCase',
				},
			],
		},
	],
})

run({
	rule,
	testDirectory: __dirname,
})
