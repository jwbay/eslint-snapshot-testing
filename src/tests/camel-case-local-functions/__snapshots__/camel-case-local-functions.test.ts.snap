// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should lint correctly for camel-case-local-functions 1`] = `
"[33mwarning[39m: [1mlocal function names should be camelCase[22m [2m(camel-case-local-functions)[22m at [32mcamel-case-local-functions.fixture:4:1[39m:
  2 | function shouldPass() {}
  3 | 
> 4 | ShouldFail();
    | ^
  5 | shouldPass();
  6 | 
  7 | ShouldFail();


[33mwarning[39m: [1mlocal function names should be camelCase[22m [2m(camel-case-local-functions)[22m at [32mcamel-case-local-functions.fixture:7:1[39m:
  5 | shouldPass();
  6 | 
> 7 | ShouldFail();
    | ^


[31m[1m2 errors found.[22m[39m"
`;
